apm-agent-plugins/apm-jms-plugin/src/main/java/co/elastic/apm/agent/jms/JmsInstrumentationHelperImpl.java

```diff
@@ -86,4 +109,49 @@ public Span startJmsProducerSpan(Destination destination, Message message) {
         }
         return span;
     }
+
+    @VisibleForAdvice
+    @Override
+    @Nullable
+    public MessageListener wrapLambda(@Nullable MessageListener listener) {
+        // the name check also prevents from wrapping twice
+        if (listener != null && listener.getClass().getName().contains("/")) {
+            return messageListenerWrapperObjectPool.createInstance().wrap(listener);
+        }
+        return listener;
+    }
+
+    private void recycle(MessageListenerWrapper messageListenerWrapper) {
+        messageListenerWrapperObjectPool.recycle(messageListenerWrapper);
+    }
+
+    public class MessageListenerWrapper implements MessageListener, Recyclable {
+
+        @Nullable
+        private MessageListener delegate;
+
+        MessageListenerWrapper() {
+        }
+
+        MessageListenerWrapper wrap(MessageListener delegate) {
+            this.delegate = delegate;
+            return this;
+        }
+
+        @Override
+        public void onMessage(Message message) {
+            try {
+                if (delegate != null) {
+                    delegate.onMessage(message);
+                }
+            } finally {
+                JmsInstrumentationHelperImpl.this.recycle(this);
```

Comment:
couldn't the same listener be invoked multiple times?

Reply:
Of course it can 🤦‍♂ 
Good catch!!
The other option of pooling them is to instrument `MessageConsumer#close` as well, and in the `setMessageListener` being able to identify already used Listeners, which means I will need to cache them in a `MessageConsumer`-->`MessageListenerWrapper` map..
WDYT?

Reply:
I mean- maybe the extra risk and complication doesn't worth the performance gain 🤔

Reply:
Discussed and decided not to pool those atm. We don't know whether the common use case is long-lived or short-lived `MessageListener`s. We can revisit if we see it has a real impact in some use cases.